src/main.rs

```diff
@@ -1,58 +1,98 @@
-    if let Some(index) = merkle[0].iter().position(|h| *h == hash) {
-        Some(index)
-    } else {
-        None
-    }
+#[derive(Debug)]
+struct MerkleTree {
+    levels: Vec<Vec<String>>
 }
 
-/// Generates a proof that a certain element belongs to the tree, if present.
-fn generate_proof(merkle: &MerkleTree, element: u32) -> Option<(Vec<String>, usize)> {
-    let Some( leaf_index) = leaf_index_for_element(merkle, element) else {
-        return None
-    };
-    let mut proof: Vec<String> = vec![];
-
-    let mut index = leaf_index;
-    for level in merkle.iter().take(merkle.len() - 1) {
-        if index % 2 == 0 {
-            proof.push(level[index+1].clone())
-        } else {
-            proof.push(level[index-1].clone())
+impl MerkleTree {
+    /// Creates a Merkle Tree given a vector of u32
+    fn create_from_values(initial_vals: &[u32]) -> Self {
+        let initial = create_initial_level(initial_vals);
+        let mut actual: Box<Vec<String>> = Box::new(initial);
+        let mut merkle = MerkleTree {
+            levels: vec![],
+        };
+        merkle.levels.push(*actual.clone());
+        while actual.len() > 1 {
+            actual = Box::new(calculate_next_level(&actual));
+            merkle.levels.push(*actual.clone());
         }
-        index = index / 2;
+        merkle
     }
 
-    Some((proof, leaf_index))
-}
+    /// Returns the height of the tree; including the root
+    fn height(&self) -> usize {
+        self.levels.len()
+    }
 
-/// Given an element, it's index and a proof verifies it against the root of a Merkle Tree
-fn verify_proof(element: u32, index: usize, proof: &[String], root: &String) -> bool {
-    let mut hash = hash_one(element);
-    let mut index = index;
+    /// Returns the leafs for the tree
+    fn leafs(&self) -> &Vec<String> {
+        &self.levels[0]
+    }
 
-    for p in proof.iter() {
-        hash = if index % 2 == 0 {
-            hash_multiple(vec![hash.clone(), p.clone()])
+    /// Returns the root hash of the tree
+    fn root(&self) -> Option<&String> {
+        if let Some(root_level) = self.levels.last() {
+            Some(&root_level[0])
         } else {
-            hash_multiple(vec![p.clone(), hash.clone()])
+            None
+        }
+    }
+
+    /// Returns the index in the leaf array for a given element, if present.
+    fn leaf_index_for_element(&self, element: u32) -> Option<usize> {
+        let hash = hash_one(element);
+        self.leafs().iter().position(|h| *h == hash)
+    }
+
+    /// Checks if the given hash is present at any level of the Merkle Tree
+    fn contains_hash(&self, hash: &String) -> bool {
+        self.levels.iter().any(|level| level.contains(hash))
+    }
+
+    /// Generates a proof that a certain element belongs to the tree, if present.
+    fn generate_proof(&self, element: u32) -> Option<(Vec<String>, usize)> {
+        let Some( leaf_index) = self.leaf_index_for_element(element) else {
+            return None
         };
```

Comment:
You can use the `?` operator so it looks more clean
`let leaf_index = self.leaf_index_for_element(element)?;`

Reply:
Done